File: chazm-model/src/main/java/runtimemodels/chazm/model/function/DefaultEffectiveness.java
package runtimemodels.chazm.model.function;

import runtimemodels.chazm.api.Organization;
import runtimemodels.chazm.api.function.Effectiveness;
import runtimemodels.chazm.api.relation.Assignment;

import javax.inject.Singleton;
import javax.validation.constraints.NotNull;
import java.util.Collection;

@Singleton
class DefaultEffectiveness implements Effectiveness {

    @Override
    public double compute(@NotNull final Organization organization, @NotNull final Collection<Assignment> assignments) {
        return assignments
                .parallelStream()
                .mapToDouble(
                        m -> organization.getGoodness(m.getRole().getId()).compute(organization, m.getAgent(), m.getRole(), m.getGoal(),
                                organization.getAssignments())).sum();
    }

}

File: chazm-model/src/main/java/runtimemodels/chazm/model/function/DefaultEffectiveness.kt
package runtimemodels.chazm.model.function

import runtimemodels.chazm.api.Organization
import runtimemodels.chazm.api.function.Effectiveness
import runtimemodels.chazm.api.relation.Assignment

import javax.inject.Singleton

@Singleton
internal open class DefaultEffectiveness : Effectiveness {

    override fun compute(organization: Organization, assignments: Collection<Assignment>): Double {
        return assignments
            .parallelStream()
            .mapToDouble { m ->
                organization.getGoodness(m.role.id).compute(organization, m.agent, m.role, m.goal,
                    organization.assignments)
            }.sum()
    }

}
